### PR TITLE
Debit card Expenses Support

### PR DESCRIPTION
**Pull request by labhvam5**

### CODE REVIEW

File: src/app/qbo/settings/general-mappings/general-mappings.component.ts [hunk 1]
```diff
@@ -250,6 +263,7 @@ export class GeneralMappingsComponent implements OnInit {
       that.billPaymentAccounts = responses.BANK_ACCOUNT;
       that.taxCodes = responses.TAX_CODE;
       that.qboExpenseAccounts = [ ...responses.BANK_ACCOUNT, ...responses.CREDIT_CARD_ACCOUNT];
+      that.debitCardExpenseAccounts = [ ...responses.BANK_ACCOUNT, ...responses.CREDIT_CARD_ACCOUNT];
```
Comment (Shwetabhk) [inline src/app/qbo/settings/general-mappings/general-mappings.component.ts:266]: why are we adding `responses.CREDIT_CARD_ACCOUNTS`?

Comment (labhvam5) [reply src/app/qbo/settings/general-mappings/general-mappings.component.ts:266]: will remove this off

File: src/app/qbo/expense-groups/expense-groups.component.ts [hunk 2]
```diff
@@ -164,6 +164,9 @@ export class ExpenseGroupsComponent implements OnInit, OnDestroy {
         } else if (expenseGroup.fund_source === 'CCC' && expenseGroup.response_logs.Purchase.PaymentType === 'CreditCard' && expenseGroup.response_logs.Purchase.Credit) {
           exportType = 'Credit Card Credit';
           exportRedirection = 'creditcardcredit';
+        } else if (expenseGroup.fund_source === 'CCC' && expenseGroup.response_logs.Purchase.PaymentType !== 'CreditCard') {
```
Comment (ashwin1111) [inline src/app/qbo/expense-groups/expense-groups.component.ts:167]: If you could add the same `response_logs`, it would be much easier to review the PR.

Comment (ashwin1111) [reply src/app/qbo/expense-groups/expense-groups.component.ts:167]: What will be the value of `expenseGroup.response_logs.Purchase.PaymentType` here?

File: src/app/qbo/settings/general-mappings/general-mappings.component.html [hunk 3]
```diff
@@ -60,12 +60,12 @@
       </mat-form-field>
       <mandatory-error-message listName="bank account" *ngIf="form.controls.bankAccounts.touched && !form.controls.bankAccounts.valid"></mandatory-error-message>
 
-      <div class="mat-body-strong" *ngIf="generalSettings.corporate_credit_card_expenses_object && generalSettings.corporate_credit_card_expenses_object !== 'BILL'">
+      <div class="mat-body-strong" *ngIf="generalSettings.corporate_credit_card_expenses_object === 'CREDIT CARD PURCHASE' && generalSettings.corporate_credit_card_expenses_object !== 'BILL'">
```
Comment (ashwin1111) [inline src/app/qbo/settings/general-mappings/general-mappings.component.html:63]: Don't we need it for Journal Entries?

File: src/app/qbo/settings/general-mappings/general-mappings.component.html [hunk 4]
```diff
@@ -60,12 +60,12 @@
...
         Default Credit Card Account
         <mandatory-field *ngIf="isFieldMandatory('cccAccounts')"></mandatory-field>
       </div>
       <mat-form-field floatLabel="always" appearance="outline" required
-      *ngIf="generalSettings.corporate_credit_card_expenses_object && generalSettings.corporate_credit_card_expenses_object !== 'BILL'">
+      *ngIf="generalSettings.corporate_credit_card_expenses_object === 'CREDIT CARD PURCHASE' && generalSettings.corporate_credit_card_expenses_object !== 'BILL'">
```
Comment (ashwin1111) [inline src/app/qbo/settings/general-mappings/general-mappings.component.html:68]: Don't we need it for Journal Entries?

File: src/app/qbo/settings/general-mappings/general-mappings.component.ts [hunk 5]
```diff
@@ -126,7 +133,7 @@ export class GeneralMappingsComponent implements OnInit {
       that.form.controls.bankAccounts.setValidators(Validators.required);
     }
 
-    if (that.generalSettings.corporate_credit_card_expenses_object && that.generalSettings.corporate_credit_card_expenses_object !== 'BILL') {
+    if (that.generalSettings.corporate_credit_card_expenses_object && that.generalSettings.corporate_credit_card_expenses_object !== 'BILL' && that.generalSettings.corporate_credit_card_expenses_object !== 'DEBIT CARD EXPENSE') {
```
Comment (ashwin1111) [inline src/app/qbo/settings/general-mappings/general-mappings.component.ts:136]: condition in template and component should match

File: src/app/qbo/settings/general-mappings/general-mappings.component.ts [hunk 6]
```diff
@@ -250,6 +263,7 @@ export class GeneralMappingsComponent implements OnInit {
       that.billPaymentAccounts = responses.BANK_ACCOUNT;
       that.taxCodes = responses.TAX_CODE;
       that.qboExpenseAccounts = [ ...responses.BANK_ACCOUNT, ...responses.CREDIT_CARD_ACCOUNT];
+      that.debitCardExpenseAccounts = [ ...responses.BANK_ACCOUNT];
```
Comment (ashwin1111) [inline src/app/qbo/settings/general-mappings/general-mappings.component.ts:266]: why not just do `responses.BANK_ACCOUNT`?

File: src/app/qbo/settings/general-mappings/general-mappings.component.ts [hunk 7]
```diff
@@ -216,7 +229,7 @@ export class GeneralMappingsComponent implements OnInit {
     attributes.push('CREDIT_CARD_ACCOUNT');
   }
 
-  if (that.generalSettings.reimbursable_expenses_object === 'EXPENSE') {
+  if (that.generalSettings.reimbursable_expenses_object === 'EXPENSE' || that.generalSettings.corporate_credit_card_expenses_object === 'DEBIT CARD EXPENSE') {
```
Comment (ashwin1111) [inline src/app/qbo/settings/general-mappings/general-mappings.component.ts:232]: I see just `BANK_ACCOUNT` is needed for `DEBIT CARD EXPENSE`. Why are we adding `CREDIT_CARD_ACCOUNT` as well?

File: src/app/qbo/expense-groups/expense-groups.component.ts (hunk 2 above)
Comment (Sravanksk) [reply src/app/qbo/expense-groups/expense-groups.component.ts:167]: `expenseGroup.response_logs.Purchase.PaymentType = 'Debit'`
Labhvam Please check this once in expense_groups table

File: src/app/qbo/settings/general-mappings/general-mappings.component.ts (hunk 5 above)
Comment (labhvam5) [reply src/app/qbo/settings/general-mappings/general-mappings.component.ts:136]: have added the same thing to the HTML template this should take care of the Journal entries part

File: src/app/qbo/expense-groups/expense-groups.component.ts (hunk 2 above)
Comment (labhvam5) [reply src/app/qbo/expense-groups/expense-groups.component.ts:167]: have added the the PaymentType  === 'Cash'

File: src/app/qbo/settings/general-mappings/general-mappings.component.html (hunk 4 above)
Comment (labhvam5) [reply src/app/qbo/settings/general-mappings/general-mappings.component.html:68]: have added this extra case for !== 'DEBIT CARD EXPENSE'
`if (that.generalSettings.corporate_credit_card_expenses_object && that.generalSettings.corporate_credit_card_expenses_object !== 'BILL' && that.generalSettings.corporate_credit_card_expenses_object !== 'DEBIT CARD EXPENSE') {
 `

File: src/app/qbo/settings/general-mappings/general-mappings.component.html (hunk 3 above)
Comment (labhvam5) [reply src/app/qbo/settings/general-mappings/general-mappings.component.html:63]: have added this extra case for !== 'DEBIT CARD EXPENSE'
if (that.generalSettings.corporate_credit_card_expenses_object && that.generalSettings.corporate_credit_card_expenses_object !== 'BILL' && that.generalSettings.corporate_credit_card_expenses_object !== 'DEBIT CARD EXPENSE') {

File: src/app/qbo/expense-groups/expense-groups.component.ts (hunk 2 above)
Comment (ashwin1111) [reply src/app/qbo/expense-groups/expense-groups.component.ts:167]: sorry, where is this added?

File: src/app/qbo/settings/general-mappings/general-mappings.component.html (hunk 3 above)
Comment (ashwin1111) [reply src/app/qbo/settings/general-mappings/general-mappings.component.html:63]: sorry, where is this added?

File: src/app/qbo/settings/general-mappings/general-mappings.component.html (hunk 4 above)
Comment (ashwin1111) [reply src/app/qbo/settings/general-mappings/general-mappings.component.html:68]: sorry, where is this added?

File: src/app/qbo/settings/general-mappings/general-mappings.component.ts (hunk 5 above)
Comment (ashwin1111) [reply src/app/qbo/settings/general-mappings/general-mappings.component.ts:136]: Whatever the condition is, it should be same in template and component

File: src/app/qbo/settings/general-mappings/general-mappings.component.ts (hunk 7 above)
Comment (ashwin1111) [reply src/app/qbo/settings/general-mappings/general-mappings.component.ts:232]: this is not fixed yet

**Review by Sravanksk — APPROVED** [review]

**Pull request merged**